### DISCUSSION TITLE
Determine the duration of an mp3 file

### DISCUSSION THREAD

**Discussion by SzymonWojd** [pschatzmann/arduino-audio-tools#1529] [Q&A] [answered]
Hi, is there any way to determine the duration in seconds of an mp3 file using this library? Also, I would be grateful if you could help me with checking how much time is left to the end of a playing file. I am building A2DP audio player which streams mp3 files and this information would be really helpful for me.

**Comment by pschatzmann** [comment, marked as answer]
For the first question: [google is your friend](https://stackoverflow.com/questions/28706093/is-it-possible-to-know-the-duration-of-an-mp3-before-the-entire-file-is-download)!

For the second: just read the actual file position and compare it with the total file size

**Reply by SzymonWojd** [reply]
Thank you for the answer. How can I read the actual file position? Is there a function in your library that does that?

**Comment by pschatzmann** [comment]
The [AudioPlayer provides you a pointer to the actual stream.](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_player.html#a190ffdf693e3bb63f5a9f560ba0f3bf9): In your case this is a pointer to a File, so you can just cast it to (File*).

**Reply by SzymonWojd** [reply]
I wrote this lines of code:
    File *pFile = (File*)player.getStream();
    Serial.println(pFile->position());
 However when I try to run the code my ESP32 just keeps resetting. Do you know what might be a problem?

**Comment by pschatzmann** [comment]
It should be
File* pFile = (File*)player.getStream();
And before using it, you must make sure that it is not nullptr!

**Reply by SzymonWojd** [reply]
I have been trying to figure out why my code is not working (problem with resetting esp32 still occurs (on serial monitor i get information that it is Guru Meditation Error: Core  1 panic'ed)). This is my code: 
if(playerI2S.getStream() != nullptr){
      File* pFile = (File*)player.getStream();
      Serial.println(pFile->position());
    }
   For me code this line was a problem: Serial.println(pFile->position()); and to be specific this function: pFile->position() even if I changed it to something like pFile->size() it still didn't feel like working...

**Reply by pschatzmann** [reply]
Which example are you using ? 
Please confirm that you are using the SD library...

**Reply by SzymonWojd** [reply]
I am using player-sdfat-a2dp example and yes i am using SD library.

**Reply by pschatzmann** [reply]
No. you are using the SDFAT library and not the SD library! 
So you need to use the data types and functionality provided by that library which would be File32, I think!

**Reply by varungujjar** [reply]
Hi, I am working on a DIY project using a PCM 5102A DAC, tried the above and I am able to get the position of the file, but in order to calculate the duration the formula requires to have a bitrate, is there a way to get the bitrate of the MP3 being played ? or is there any other method for calculating the same ?

**Reply by pschatzmann** [reply]
The MP3DecoderHelix provides the audioInfoEx() method which return a MP3FrameInfo struct that contains the bitrate.

**Reply by varungujjar** [reply]
Thankyou @pschatzmann  for such a quick response, sorry but I am little new to this and I am getting the following
44,100 Khz / 16 Bit / 2 Channels and calculate  using a forumla  44100*16 *2 = 1, 411 Kbps, but  is there a way to get the actual bit rate such as 128kbps, 192Kbps etc. ?

**Reply by hash6iron** [reply]
Hello @pschatzmann

Yes, the bitrate could be consulted in audioInfoEx() from decoder, but I've observed that VBR MP3 don't play correctly when bitrate is consulted each time. Is there any solution to calculate duration for a variable bit rate MP3 avoiding playing issues?

Thanks for all.

**Reply by pschatzmann** [reply]
Why would you do something like this ? You would do this only once to get a rough first estimate (if ever).
You can use the MeasuringStream to estimate the total runime or the time left using the approach described in my first answer:
- calling begin records the start time
- putting it into the chain or calling setProcessedBytes() with the actual file position records the progress
- estimatedOpenTimeFor(file size) gives the remaining time
- estimatedTotalTimeFor(file size) give the updated total time

**Reply by pschatzmann** [reply]
I have added an [example to the documentation](https://github.com/pschatzmann/arduino-audio-tools/wiki/Measuring-the-Data-Thruput)
